### PR TITLE
CircleCI POC - Main branch

### PR DESCRIPTION
**Pull request by grena**
Main PR for the Circle CI project.
Please base all your PR on this one and DO NOT REMOVE this branch.

### CODE REVIEW

File: .circleci/behat.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+tests/legacy/features/pim/enrichment/product/completeness/remove_completeness.feature:25
```
Comment (grena) [inline .circleci/behat.txt]: To remove

File: .circleci/phpunit.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/SuccessLargeAndOrderedListProductEndToEnd.php
```
Comment (grena) [inline .circleci/phpunit.txt]: To remove

File: toto.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+tests/back/Pim/Enrichment/Specification/Component/Product/Builder/EntityWithValuesBuilderSpec.php
```
Comment (grena) [inline toto.txt]: To remove toto

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+                name: PhpSpec
+                command:
+                    docker-compose exec -T fpm vendor/bin/phpspec run --format=junit > var/tests/phpspec/specs.xml
+          - run:
+                name: PhpCsFixer
+                command: docker-compose exec -T fpm vendor/bin/php-cs-fixer fix --diff --dry-run --config=.php_cs.php
+          - run:
+                name: PhpCoupling Detector
+                command: |
+                    set -e
```
Comment (jjanvier) [inline .circleci/config.yml]: make coupling

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+  test_acceptance_back:
+      machine:
+          enabled: true
+          docker_layer_caching: true
+      steps:
+          - attach_workspace:
+                at: ~/
+          - run:
+              name: Change owner on project dir (default user = circleci (1001) and docker needs uid 1000)
```
Comment (jjanvier) [inline .circleci/config.yml]: why do we need that here?
it's already done in the "build" step

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+          docker_layer_caching: true
+      steps:
+          - attach_workspace:
+                at: ~/
+          - run:
+                name: Change owner on project dir (default user = circleci (1001) and docker needs uid 1000)
+                command: sudo chown -R 1000:1000 ../project
+          - run:
+                name: Start containers
+                command:
```
Comment (jjanvier) [inline .circleci/config.yml]: one line

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+                    docker-compose up -d fpm
+          - run:
+                name: PhpSpec
+                command:
```
Comment (jjanvier) [inline .circleci/config.yml]: one line

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+          docker_layer_caching: true
+      steps:
+          - attach_workspace:
+                at: ~/
+          - run:
+              name: Change owner on project dir (default user = circleci (1001) and docker needs uid 1000)
+              command: sudo chown -R 1000:1000 ../project
+          - run:
+                name: Start containers
+                command: |
```
Comment (jjanvier) [inline .circleci/config.yml]: one line

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+      - checkout
+      - run:
+          name: Copy env file
+          command: cp .env.dist .env
+      - run:
+          name: Copy docker-compose.override.yml.dist
+          command: cp .circleci/docker-compose.override.yml.dist docker-compose.override.yml
+      - run:
+          name: Setup tests results folder and log folder
+          command: mkdir -p var/tests/phpunit var/tests/behat var/tests/phpspec var/tests/csfixer var/logs
```
Comment (jjanvier) [inline .circleci/config.yml]: should we separate phpunit?

no junit for back acceptance?
no junit for frontend tests?

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+      - run:
+          name: Change owner on project dir (default user = circleci (1001) and docker needs uid 1000)
+          command: sudo chown -R 1000:1000 ../project
+      - run:
+          name: Start containers
+          command: docker-compose up -d
+      - restore_cache:
+          name: Restore cache - node_modules
+          keys:
+            - node-modules-v7-{{ checksum "yarn.lock" }}
```
Comment (jjanvier) [inline .circleci/config.yml]: ideally, env variable

File: .circleci/config.yml [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+      - restore_cache:
+          name: Restore cache - vendor
+          keys:
+            - vendor-v7-{{ checksum "composer.lock" }}
```
Comment (jjanvier) [inline .circleci/config.yml]: ideally, env variable

File: .circleci/config.yml [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+      - run:
+            name: Pim yarn test installation
+            command: docker-compose run --rm node yarn run webpack-test
+      - run:
+            name: Change owner on project dir after installing when there is no cache
+            command: sudo chmod -R 777 ../project
+      - save_cache:
+          paths:
+            - ./node_modules
+          key: node-modules-v7-{{ checksum "yarn.lock" }}
```
Comment (jjanvier) [inline .circleci/config.yml]: ideally, env variable

File: .circleci/config.yml [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+      - save_cache:
+          paths:
+            - ./vendor
+          key: vendor-v7-{{ checksum "composer.lock" }}
```
Comment (jjanvier) [inline .circleci/config.yml]: ideally, env variable

File: .circleci/config.yml [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+                command: docker-compose exec -T fpm vendor/bin/phpstan analyse src/Akeneo/Pim -l 0
+          - run:
+                name: Check Pullup
+                command: docker-compose exec -T fpm bin/check-pullup
+          - store_test_results:
+                path: var/tests
+          - store_artifacts:
+                path: var/tests
+
+  test_acceptance_back:
```
Comment (jjanvier) [inline .circleci/config.yml]: useless to //

File: .circleci/config.yml [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+            command: docker-compose run --rm node yarn run lint
+        - run:
+            name: Yarn Unit
+            command: docker-compose run --rm node yarn run unit
+        - run:
+            name: Front acceptance tests
+            command: |
+              docker-compose run --rm node yarn run webpack-test
+              MAX_RANDOM_LATENCY_MS=100 docker-compose run --rm node yarn run acceptance ./tests/features
+
```
Comment (jjanvier) [inline .circleci/config.yml]: missing front integration https://github.com/akeneo/pim-community-dev/blob/master/.ci/Jenkinsfile#L150

File: .circleci/config.yml [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+          - run:
+                name: Change owner on project dir in order to archive the project into the workspace
+                command: sudo chown -R 1000:1000 ../project
+          - run:
+                name: Start containers
+                command: docker-compose up -d
+          - run:
+                name: Install database
+                command: |
+                    sleep 25 # wait mysql to be up
```
Comment (jjanvier) [inline .circleci/config.yml]: script to wait for MySQL to be up

File: .circleci/config.yml [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
```
Comment (jjanvier) [inline .circleci/config.yml]: script to wait for MySQL to be up

File: .circleci/config.yml [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+                    docker-compose exec fpm bin/console --env=behat pim:install:db
+          - run:
+                name: PHPunit End to end
+                command: |
+                    TESTFILES=$(circleci tests glob "tests/back/**/*EndToEnd.php" | circleci tests split)
+                    .circleci/run_phpunit.sh $TESTFILES
+          - store_test_results:
+                path: var/tests/phpunit
```
Comment (jjanvier) [inline .circleci/config.yml]: split?

File: .circleci/config.yml [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+                command: |
+                    sleep 25 # wait mysql to be up
+                    docker-compose exec fpm bin/console --env=behat pim:install:db
+          - run:
+                name: PHPunit Integration
+                command: |
+                    TESTFILES=$(circleci tests glob "tests/back/**/*Integration.php" | circleci tests split)
+                    .circleci/run_phpunit.sh $TESTFILES
+          - store_test_results:
+                path: var/tests/phpunit
```
Comment (jjanvier) [inline .circleci/config.yml]: split?

**Pull request merged**